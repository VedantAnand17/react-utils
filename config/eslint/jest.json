{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}