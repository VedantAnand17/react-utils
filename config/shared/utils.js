/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
