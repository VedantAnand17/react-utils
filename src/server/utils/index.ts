// Aggregates utility modules together.

export * as errors from './errors';
