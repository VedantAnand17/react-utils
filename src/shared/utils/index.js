import path from 'path';

import config from './config';
import * as isomorphy from './isomorphy';
import * as redux from './redux';
import * as webpack from './webpack';

const juUrl = module.webpackPolyfill ? './shared/utils/jest'
  : path.resolve(__dirname, './jest');

const JU = isomorphy.isServerSide() ? webpack.requireWeak(juUrl) : null;

export {
  config,
  isomorphy,
  JU,
  redux,
  webpack,
};
