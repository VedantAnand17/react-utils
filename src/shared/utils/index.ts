import themedImpl, {
  type Theme,
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

import type * as JuT from './jest';

export {
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

// Note: it should be done this way, as in some environments
// "process" might not exist, and process.env.NODE_CONFIG_ENV
// not injected by Webpack.
let NODE_CONFIG_ENV;
try {
  NODE_CONFIG_ENV = process.env.NODE_CONFIG_ENV;
} catch { /* noop */ }

const env = NODE_CONFIG_ENV || process.env.NODE_ENV;

const JU: typeof JuT | null = env !== 'production'
  ? webpack.requireWeak('./jest', __dirname) as typeof JuT | null
  : null;

export {
  type Theme,
  config,
  isomorphy,
  JU,
  themed,
  ThemeProvider,
  time,
  webpack,
};
