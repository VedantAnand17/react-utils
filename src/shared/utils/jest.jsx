/**
 * Utilities for Jest tests.
 */

/* NOTE: At the moment here is no intention to use this module for runtime
 * execution, only for tests; thus we depend on Jest environment and dev
 * dependencies. */

/* TODO: Revise.  */

/* global expect, jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';

import mockdate from 'mockdate';
import PT from 'prop-types';
import { render as newRender, unmountComponentAtNode } from 'react-dom';
import TU from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/* NEW TESTING METHODS */

export { act } from 'react-dom/test-utils';

const originalProcessVersions = process.versions;

export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: null });
}

export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  newRender(scene, res);
  return res;
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}

/* OLD STUFF BELOW THIS MARK */

/**
 * Just an alias for TU.findRenderedDOMComponentWithClass(..).
 * @param {Object} dom
 * @param {String} className
 * @return {Object}
 */
export function findInDomByClass(dom, className) {
  return TU.findRenderedDOMComponentWithClass(dom, className);
}

/**
 * Just an alias for TU.scryRenderedDOMComponentsWithClass(..).
 * @param {Object} dom
 * @param {Stirng} className
 * @return {Array}
 */
export function findInDomManyByClass(dom, className) {
  return TU.scryRenderedDOMComponentsWithClass(dom, className);
}

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
export function render(component) {
  return Renderer.create(component).toJSON();
}

/* The Wrapper is necessary for the "renderDom(..)" function, because
 * the "renderIntoDocument(..)" function from "react-dom/test-utils" works
 * only with state components, so we have to wrap our ReactJS components into
 * such Wrapper. */
class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    const { children } = this.props;
    return children;
  }
}

Wrapper.propTypes = {
  children: PT.node.isRequired,
};

/**
 * Renders given component into DOM, using "react-dom/test-utils".
 * @param {Object} component ReactJS component to render.
 * @return {Object} Rendered DOM.
 */
export function renderDom(component) {
  return TU.renderIntoDocument((
    <Wrapper>
      {component}
    </Wrapper>
  ));
}

/**
 * Auxiliary wrapper around ReactJS Shallow Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Renders provided ReactJS component with ReactJS Shallow Test Renderer,
 * creates/tests its snapshot, and returns the render result.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of shallow render.
 */
export function shallowSnapshot(component) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
