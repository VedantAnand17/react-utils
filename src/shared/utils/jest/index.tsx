/* global expect, jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import mockdate from 'mockdate';
import { type ReactElement, type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import TU, { act } from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import { type ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * An alias for [act(..)](https://reactjs.org/docs/test-utils.html#act)
 * from `react-dom/test-utils`.
 * @param {function} action
 */
export { act };

declare global {
  // eslint-disable-next-line no-var, vars-on-top
  var IS_REACT_ACT_ENVIRONMENT: boolean | undefined;
}

global.IS_REACT_ACT_ENVIRONMENT = true;

const originalProcessVersions = process.versions;

/**
 * Tricks **react-utils** into thinking the test is running within client-side
 * (browser) environment.
 */
export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: null });
}

/**
 * Reverts the effect of {@link module:JU.mockClientSide mockClientSide(..)}.
 */
export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID, or better said it determenistically transforms given
 * `seed` number into a UUID-formatted string.
 * @param {number} seed
 * @return {string}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 * be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

type MountedSceneT = HTMLElement & {
  destroy: () => void;
};

/**
 * Mounts `scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return Created container DOM element with destroy() function
 *  attached.
 */
export function mount(scene: ReactNode): MountedSceneT {
  let root: Root;
  const element = document.createElement('div');
  document.body.appendChild(element);

  const res: MountedSceneT = (element as unknown) as MountedSceneT;
  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  act(() => {
    root = createRoot(res);
    root.render(scene);
  });
  return res;
}

/* OLD STUFF BELOW THIS MARK */

/**
 * Renders provided ReactJS component into JSON representation of the component
 * tree, using [`react-test-renderer`](https://www.npmjs.com/package/react-test-renderer).
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of the rendered tree.
 * @example
 * import { JU } from '@dr.pogodin/react-utils';
 * console.log(JU.render(<div>Example</div>));
 */
export function render(component: ReactElement) {
  return Renderer.create(component).toJSON();
}

/**
 * Generates a shallow render of given ReactJS component, using
 * [react-test-renderer/shallow](https://reactjs.org/docs/shallow-renderer.html)
 * and returns the result.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component: ReactElement) {
  const renderer: ShallowRenderer = createRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Makes a shallow snapshot test of the given ReactJS component, and also
 * returns JSON representation of the rendered component tree. Under the hood
 * it uses {@link module:JU.shallowRender shallowRender(..)} to generate
 * the render, then executes `expect(RENDER_RESULT).toMatchSnapshot()`,
 * and finally returns the `RENDER_RESULT` to the caller.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON representation of shallow render.
 */
export function shallowSnapshot(component: ReactElement) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Makes snapshot test of the given ReactJS component, and also returns JSON
 * representation of the rendered component tree. Under the hood, it uses
 * {@link module:JU.render render(..)} to render it, then executes
 * `expect(RENDER_RESULT).toMatchSnapshot()`, and then returns `RENDER_RESULT`.
 * @param {object} component ReactJS component to render.
 * @return {object} JSON render of the component.
 */
export function snapshot(component: ReactElement) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
