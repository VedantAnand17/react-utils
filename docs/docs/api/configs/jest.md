# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
