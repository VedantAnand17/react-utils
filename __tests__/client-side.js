/** @jest-environment jsdom */

import mockdate from 'mockdate';
import { mockClientSide } from 'utils/jest';

mockdate.set('2020-04-19Z');

// TODO: This was disabled after upgrade of axios to 1.0.0, as it uses not
// compiled ES imports, and thus requires client side to use webpack build
// environment. However, current implementation of E2eSsrEnv has misc issues
// to make it work: SASS fails if we can't set correct context needed to resolve
// SCSS imports.
test.skip('Export at client side', () => {
  mockClientSide();
  const lib = require('../src');
  expect(lib).toMatchSnapshot();
  expect({ ...lib._ }).toMatchSnapshot();
  expect({ ...lib.api }).toMatchSnapshot();
  expect(Object.keys(lib.time)).toMatchSnapshot();
});
