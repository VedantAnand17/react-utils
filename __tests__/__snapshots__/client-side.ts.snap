// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export at client side 1`] = `
{
  "Barrier": [Function],
  "BaseButton": [Function],
  "BaseModal": [Function],
  "Button": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Checkbox": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "CustomDropdown": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Dropdown": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Emitter": [Function],
  "GlobalStateProvider": [Function],
  "Input": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Link": [Function],
  "MetaTags": [Function],
  "Modal": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "NavLink": [Function],
  "PageLayout": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Semaphore": [Function],
  "Switch": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "TextArea": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "ThemeProvider": [Function],
  "Throbber": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "WithTooltip": {
    "$$typeof": Symbol(react.forward_ref),
    "PLACEMENTS": {
      "ABOVE_CURSOR": "ABOVE_CURSOR",
      "ABOVE_ELEMENT": "ABOVE_ELEMENT",
      "BELOW_CURSOR": "BELOW_CURSOR",
      "BELOW_ELEMENT": "BELOW_ELEMENT",
    },
    "render": [Function],
  },
  "YouTubeVideo": {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "client": [Function],
  "config": {
    "CSRF": undefined,
    "MOCK_CONFIG_KEY": "Mock Config Value",
  },
  "getGlobalState": [Function],
  "getSsrContext": [Function],
  "isomorphy": {
    "IS_CLIENT_SIDE": true,
    "IS_SERVER_SIDE": false,
    "buildTimestamp": [Function],
    "getBuildInfo": [Function],
    "isDevBuild": [Function],
    "isProdBuild": [Function],
  },
  "newAsyncDataEnvelope": [Function],
  "server": null,
  "splitComponent": [Function],
  "themed": [Function],
  "time": [Function],
  "useAsyncCollection": [Function],
  "useAsyncData": [Function],
  "useGlobalState": [Function],
  "webpack": {
    "requireWeak": [Function],
    "resolveWeak": [Function],
  },
  "withGlobalStateType": [Function],
  "withRetries": [Function],
}
`;

exports[`Export at client side 2`] = `{}`;

exports[`Export at client side 3`] = `
[
  "extend",
  "locale",
  "isDayjs",
  "unix",
  "en",
  "Ls",
  "p",
  "DAY_MS",
  "HOUR_MS",
  "MIN_MS",
  "SEC_MS",
  "YEAR_MS",
  "now",
  "timer",
  "useCurrent",
  "useTimezoneOffset",
]
`;
